// zrender/lib/core/vector.js
"use strict";
exports.__esModule = true;
function create(x, y) {
    if (x == null) {
        x = 0;
    }
    if (y == null) {
        y = 0;
    }
    return [x, y];
}
exports.create = create;
function copy(out, v) {
    out[0] = v[0];
    out[1] = v[1];
    return out;
}
exports.copy = copy;
function clone(v) {
    return [v[0], v[1]];
}
exports.clone = clone;
function set(out, a, b) {
    out[0] = a;
    out[1] = b;
    return out;
}
exports.set = set;
function add(out, v1, v2) {
    out[0] = v1[0] + v2[0];
    out[1] = v1[1] + v2[1];
    return out;
}
exports.add = add;
function scaleAndAdd(out, v1, v2, a) {
    out[0] = v1[0] + v2[0] * a;
    out[1] = v1[1] + v2[1] * a;
    return out;
}
exports.scaleAndAdd = scaleAndAdd;
function sub(out, v1, v2) {
    out[0] = v1[0] - v2[0];
    out[1] = v1[1] - v2[1];
    return out;
}
exports.sub = sub;
function len(v) {
    return Math.sqrt(lenSquare(v));
}
exports.len = len;
exports.length = len;
function lenSquare(v) {
    return v[0] * v[0] + v[1] * v[1];
}
exports.lenSquare = lenSquare;
exports.lengthSquare = lenSquare;
function mul(out, v1, v2) {
    out[0] = v1[0] * v2[0];
    out[1] = v1[1] * v2[1];
    return out;
}
exports.mul = mul;
function div(out, v1, v2) {
    out[0] = v1[0] / v2[0];
    out[1] = v1[1] / v2[1];
    return out;
}
exports.div = div;
function dot(v1, v2) {
    return v1[0] * v2[0] + v1[1] * v2[1];
}
exports.dot = dot;
function scale(out, v, s) {
    out[0] = v[0] * s;
    out[1] = v[1] * s;
    return out;
}
exports.scale = scale;
function normalize(out, v) {
    var d = len(v);
    if (d === 0) {
        out[0] = 0;
        out[1] = 0;
    }
    else {
        out[0] = v[0] / d;
        out[1] = v[1] / d;
    }
    return out;
}
exports.normalize = normalize;
function distance(v1, v2) {
    return Math.sqrt((v1[0] - v2[0]) * (v1[0] - v2[0])
        + (v1[1] - v2[1]) * (v1[1] - v2[1]));
}
exports.distance = distance;
exports.dist = distance;
function distanceSquare(v1, v2) {
    return (v1[0] - v2[0]) * (v1[0] - v2[0])
        + (v1[1] - v2[1]) * (v1[1] - v2[1]);
}
exports.distanceSquare = distanceSquare;
exports.distSquare = distanceSquare;
function negate(out, v) {
    out[0] = -v[0];
    out[1] = -v[1];
    return out;
}
exports.negate = negate;
function lerp(out, v1, v2, t) {
    out[0] = v1[0] + t * (v2[0] - v1[0]);
    out[1] = v1[1] + t * (v2[1] - v1[1]);
    return out;
}
exports.lerp = lerp;
function applyTransform(out, v, m) {
    var x = v[0];
    var y = v[1];
    out[0] = m[0] * x + m[2] * y + m[4];
    out[1] = m[1] * x + m[3] * y + m[5];
    return out;
}
exports.applyTransform = applyTransform;
function min(out, v1, v2) {
    out[0] = Math.min(v1[0], v2[0]);
    out[1] = Math.min(v1[1], v2[1]);
    return out;
}
exports.min = min;
function max(out, v1, v2) {
    out[0] = Math.max(v1[0], v2[0]);
    out[1] = Math.max(v1[1], v2[1]);
    return out;
}
exports.max = max;
